### ISSUE TITLE
FindBugs relating to synchronised blocks

### ISSUE THREAD

**Issue by Mecharyry** [novoda/download-manager#284]
### Problem
FindBugs threw up a few issues with the synchronised blocks in version 2 of the `download-manager`. 
![screen shot 2018-01-03 at 12 55 45](https://user-images.githubusercontent.com/3380092/34521423-95246be4-f085-11e7-9ff0-3b34418312ad.png)

## Solution
Investigate ways to remove these warnings and then remove the suppression.

**Comment by MechaRyRy** [comment]
Still needs to be fixed. Not sure we need wait in a loop, we are doing a simple object lock.

**Comment by zegnus** [comment]
This might be a false positive and FindBug is not finding out that there is indeed a conditional wait:

```
T thenPerform(Action<T> action) {
            if (instanceToWaitFor == null) {
                waitForLock();
            }
            return action.performAction();
        }

        private void waitForLock() {
            try {
                synchronized (lock) {
                    if (instanceToWaitFor == null) {
                        lock.wait();
                    }
                }
            } catch (InterruptedException e) {
                Logger.e(e, "Interrupted waiting for instance.");
            }
        }
```

**Comment by MechaRyRy** [comment]
Perhaps it's worth me inlining it and checking findbugs. We could verify the false positive, keeping it in the readable format if it is indeed a false positive.

**Comment by MechaRyRy** [comment]
Addressed #433
